### PR TITLE
feat(subsystembenchmarks): add PyTorch Lightning checkpoint load benchmark

### PR DESCRIPTION
**Pull request by yuxin00j**
### Overview
This PR introduces the **checkpoint read (load) benchmark suite** for PyTorch Lightning in the subsystem benchmark framework (`gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read`). 

It extends our benchmarking capabilities to measure checkpoint loading throughput across various single-node and distributed strategies, supporting both monolithic and distributed/sharded checkpoint layouts as well as cross-topology resharding.

---

### Key Changes
1. **PyTorch Lightning Read Driver (`PLCheckpointReadDriver`)**:
   * Implemented checkpoint loading logic across single-node, DDP, FSDP (sharded & full), and Fabric Model Parallel (sharded & full) strategies.
   * Proper strategy lifecycle orchestration: automatically handles distributed environment setup, device mesh creation, module parallelization hooks (`configure_model`), and `strategy.setup()`.
   * Standardized checkpoint setup step (`setup_checkpoint_for_load`) to initialize realistic model checkpoints on GCS prior to read measurements.

2. **Common Checkpoint Infrastructure (`common.py`)**:
   * Modularized `DummyModel`, `DummyDataset`, device mesh configuration, and process management (`run_split`) for reuse across both read and write benchmark drivers.
   * Added `CPUFSDPStrategy` and `CPUModelParallelStrategy` overrides to enable CPU-based distributed testing and verification.

3. **Configurator & Case Enhancements**:
   * Updated `CheckpointCase` and `CheckpointConfigurator` to support `checkpoint_read` scenarios and metrics alongside `checkpoint_write`.
   * Added explicit scenario validation (`read` vs `write`) to ensure correct throughput and latency metric reporting.

4. **Unit Tests**:
   * Added unit tests for read configuration generators and `CheckpointCase` verification.

### CODE REVIEW

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/checkpoint_case.py [hunk 1]
```diff
@@ -66,12 +76,22 @@ def run_checkpoint_case(benchmark, monitor, params, driver, *, bucket_ctx=None):
...
-            if physical_size_bytes > 0 and durations and all(durations)
+        size_for_throughput = physical_size_bytes
+        throughput = (
+            sum(size_for_throughput / d for d in durations) / len(durations)
+            if size_for_throughput > 0 and durations and all(durations)
             else 0.0
         )
+        if "read" in params.scenario:
+            benchmark.extra_info["checkpoint_read_throughput_mean_bytes_per_second"] = (
+                throughput
```
Comment (zhixiangli) [inline gcsfs/tests/perf/subsystembenchmarks/checkpointing/checkpoint_case.py:105]: For DDP, FSDP full and Model Parallel full, the throughput calculation is incorrect.

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    trainer2.strategy.setup_environment()
+    _call_configure_model(trainer2)
+    trainer2.strategy.setup(trainer2)
+    trainer2.optimizers = [model2.configure_optimizers()]
+
+    durations = []
+    for _ in range(params.rounds):
+        dist.barrier()
+        t_start = time.perf_counter()
+        trainer2.strategy.load_checkpoint(ckpt_path)
```
Comment (zhixiangli) [inline gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py]: Have you verified that the checkpoint load size is approximately 48GB for llama3.1 8B in DDP?

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/checkpoint_case.py [hunk 3]
```diff
@@ -25,9 +28,16 @@ def run_checkpoint_case(benchmark, monitor, params, driver, *, bucket_ctx=None):
         prefix = f"gs://{bucket}/checkpoint/"
         assert_fsspec_gcsfs(prefix)
 
+        driver.setup(prefix, params)
+
         window_start = time.time()
         with monitor() as m:
-            result = driver.run_save(prefix, params)
+            if hasattr(driver, "run"):
```
Comment (zhixiangli) [inline gcsfs/tests/perf/subsystembenchmarks/checkpointing/checkpoint_case.py]: is `hasattr` necessary?

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/common.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    else:
+        raise ValueError(f"Unknown strategy: {strategy_name}")
+
+
+def run_split(prefix, params, target_fn, world_size_override=None):
+    """Spawns processes to run the distributed benchmark and gathers timing."""
+    import torch.multiprocessing as mp
+
+    ctx = mp.get_context("spawn")
+    world_size = 2 if not hasattr(params, "world_size") else params.world_size
```
Comment (zhixiangli) [inline gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/common.py]: Could you review all the `hasattr` calls? I don't believe they are necessary.

**Review by zhixiangli — CHANGES_REQUESTED** [review]

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/common.py (hunk 4 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/common.py]: Done

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/checkpoint_case.py (hunk 3 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/subsystembenchmarks/checkpointing/checkpoint_case.py]: Done.

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py (hunk 2 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py]: Yes.

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/checkpoint_case.py (hunk 1 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/subsystembenchmarks/checkpointing/checkpoint_case.py:105]: Done.

File: gcsfs/tests/perf/subsystembenchmarks/_common/benchmark_publish.py [hunk 5]
```diff
@@ -43,6 +43,8 @@ def publish_resource_metrics(benchmark, monitor):
             # max_cpu is percent normalized by vCPUs; convert it back to cores.
             "cpu_usage_peak_cores": (monitor.max_cpu * monitor.vcpus / 100.0),
             "memory_usage_peak_bytes": int(monitor.max_mem),
+            "network_received_bytes": int(monitor.net_recv),
```
Comment (Yonghui-Lee) [inline gcsfs/tests/perf/subsystembenchmarks/_common/benchmark_publish.py:46]: Do we want to add them to subsystembenchmarks_schema.json?

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/checkpoint_case.py [hunk 6]
```diff
@@ -66,12 +84,32 @@ def run_checkpoint_case(benchmark, monitor, params, driver, *, bucket_ctx=None):
...
+            size_for_throughput *= params.world_size
+
+        throughput = (
+            sum(size_for_throughput / d for d in durations) / len(durations)
+            if size_for_throughput > 0 and durations and all(durations)
             else 0.0
         )
+        if "read" in params.scenario:
+            benchmark.extra_info["checkpoint_read_throughput_mean_bytes_per_second"] = (
+                throughput
```
Comment (Yonghui-Lee) [inline gcsfs/tests/perf/subsystembenchmarks/checkpointing/checkpoint_case.py:105]: We may need to query cloud monitoring to get the actual raw bytes sent by GCS over the network like macro benchmark. So we can know `checkpoint_read_bytes` and `checkpoint_read_amplification_ratio`.

**Review by Yonghui-Lee — APPROVED** [review]

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/checkpoint_case.py [hunk 7]
```diff
@@ -66,12 +84,32 @@ def run_checkpoint_case(benchmark, monitor, params, driver, *, bucket_ctx=None):
...
+        size_for_throughput = physical_size_bytes
+
+        # In DDP, FSDP full and Model Parallel full, all ranks read the monolithic checkpoint independently.
+        # The effective data transferred from storage is multiplied by the world size.
+        if "read" in params.scenario and params.strategy in (
+            "ddp",
+            "fsdp_full",
+            "model_parallel_full",
+        ):
+            size_for_throughput *= params.world_size
```
Comment (Yonghui-Lee) [inline gcsfs/tests/perf/subsystembenchmarks/checkpointing/checkpoint_case.py]: run_split clamps the real rank count to min(world_size, 8); a world_size: 16 config would report exactly 2x throughput with no error.

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        }
+
+        ckpt_path = os.path.join(prefix, "model.ckpt")
+
+        model2 = DummyModel(params)
+        model2 = model2.to(torch.bfloat16)
+        trainer2 = L.Trainer(**trainer_args)
+        trainer2.strategy.connect(model2)
+        model2.trainer = trainer2
+        trainer2.state.fn = TrainerFn.TESTING
```
Comment (Yonghui-Lee) [inline gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py]: TrainerFn.TESTING makes FSDP load_checkpoint skip optimizer-state restore, so the timed read covers only model weights while the size denominator includes the AdamW moments; read throughput reported ~3x high for the sharded cases.

https://github.com/Lightning-AI/pytorch-lightning/blob/fcef404516264798b6c582ef556eaa17e5c29bfb/src/lightning/pytorch/strategies/fsdp.py#L619

**Review by Yonghui-Lee — CHANGES_REQUESTED** [review]

File: gcsfs/tests/perf/subsystembenchmarks/_common/benchmark_publish.py (hunk 5 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/subsystembenchmarks/_common/benchmark_publish.py:46]: Done.

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/checkpoint_case.py (hunk 6 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/subsystembenchmarks/checkpointing/checkpoint_case.py:105]: Done.

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/checkpoint_case.py (hunk 7 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/subsystembenchmarks/checkpointing/checkpoint_case.py]: Done.

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py (hunk 8 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py]: Done.

File: gcsfs/tests/perf/subsystembenchmarks/dataloading/amplification.py [hunk 9]
```diff
@@ -18,6 +18,12 @@
     "dataset_read_bytes",
     "dataset_read_request_count",
     "dataset_read_amplification_ratio",
+    "checkpoint_read_bytes",
+    "checkpoint_read_request_count",
+    "checkpoint_read_amplification_ratio",
+    "checkpoint_read_bytes",
+    "checkpoint_read_request_count",
+    "checkpoint_read_amplification_ratio",
```
Comment (Yonghui-Lee) [inline gcsfs/tests/perf/subsystembenchmarks/dataloading/amplification.py]: It's duplicated

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+
+class PLCheckpointReadDriver(CheckpointDriver):
+    def setup(self, prefix, params):
+        if params.strategy in (
+            "ddp",
+            "fsdp_sharded",
+            "fsdp_full",
+            "model_parallel_full",
+            "model_parallel_sharded",
+        ):
```
Comment (Yonghui-Lee) [inline gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py]: could you help define a helper? It's used in several places.

```
def is_distributed_strategy(strategy: str) -> bool:
    return strategy != "single"
```

**Review by Yonghui-Lee — APPROVED** [review]

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py (hunk 10 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py]: Done.

File: gcsfs/tests/perf/subsystembenchmarks/dataloading/amplification.py (hunk 9 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/subsystembenchmarks/dataloading/amplification.py]: Done.

File: gcsfs/tests/perf/subsystembenchmarks/dataloading/amplification.py [hunk 11]
```diff
@@ -98,27 +101,52 @@ def enrich_csv(csv_path, project, *, client):
...
+
+        # Check if this row is for dataset or checkpoint
+        prefix = None
+        if "dataset_size_bytes" in row and row["dataset_size_bytes"]:
+            prefix = "dataset"
+        elif (
+            "checkpoint_physical_size_bytes" in row
+            and row["checkpoint_physical_size_bytes"]
+        ):
+            prefix = "checkpoint"
```
Comment (zhixiangli) [inline gcsfs/tests/perf/subsystembenchmarks/dataloading/amplification.py:113]: Would checkpoint write rows be considered candidates for read amplification?

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+        model2 = DummyModel(params)
+        model2 = model2.to(torch.bfloat16)
+        trainer2 = L.Trainer(**trainer_args)
+        trainer2.strategy.connect(model2)
+        model2.trainer = trainer2
+        trainer2.state.fn = TrainerFn.FITTING
+        trainer2.strategy.setup_environment()
+        _call_configure_model(trainer2)
+        trainer2.strategy.setup(trainer2)
+        trainer2.optimizers = [model2.configure_optimizers()]
```
Comment (zhixiangli) [inline gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py]: Will this trigger the optimizer to load again?

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+            "devices": 1,
+            "max_steps": 1,
+            "precision": "bf16-mixed",
+            "enable_checkpointing": False,
+        }
+
+        ckpt_path = os.path.join(prefix, "model.ckpt")
+
+        model2 = DummyModel(params)
+        model2 = model2.to(torch.bfloat16)
```
Comment (zhixiangli) [inline gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py]: There's duplicate identical setup logic for loading checkpoints

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+        trainer2.state.fn = TrainerFn.FITTING
+        trainer2.strategy.setup_environment()
+        _call_configure_model(trainer2)
+        trainer2.strategy.setup(trainer2)
+        trainer2.optimizers = [model2.configure_optimizers()]
+
+        durations = []
+        for _ in range(params.rounds):
+            begin = time.perf_counter()
+            trainer2.strategy.load_checkpoint(ckpt_path)
```
Comment (zhixiangli) [inline gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py]: Single-device and DDP benchmarks deserialize checkpoints without restoring them, but others do.

File: gcsfs/tests/perf/subsystembenchmarks/dataloading/amplification.py (hunk 11 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/subsystembenchmarks/dataloading/amplification.py:113]: Done.

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py (hunk 13 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py]: Done.

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py (hunk 14 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py]: Done.

File: gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py (hunk 12 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/subsystembenchmarks/checkpointing/pytorch_lightning/read/driver.py]: Done. Removed this duplicate logic as it is done in trainer.strategy.setup.

**Review by zhixiangli — APPROVED** [review]
